src/main/java/fi/okm/jod/yksilo/service/ehdotus/MahdollisuudetService.java

```diff
@@ -44,6 +49,50 @@ public class MahdollisuudetService {
       WHERE kk.kaannos_key = :lang AND k.aktiivinen IS true ) ORDER BY otsikko COLLATE  "{0}" {1}
       """;
 
+  private static final String SQL_OSAAMINEN_SUGGESTIONS_QUERY =
+      """
+      WITH mahdollisuus AS (
+        SELECT
+          k.id as id,
+        COUNT(a) as totalMatch,
+        (SELECT SIZE(j2.arvot) FROM k.jakaumat j2
+         WHERE j2.tyyppi = :koulutusJakaumaTyyppi) as totalOsaamiset,
+        'KOULUTUSMAHDOLLISUUS' as tyyppi
+      FROM Koulutusmahdollisuus k
+      JOIN k.jakaumat j
+      JOIN j.arvot a
+      WHERE k.aktiivinen IS true AND j.tyyppi = :koulutusJakaumaTyyppi
+      AND a.arvo IN :missingOsaamiset
+      GROUP BY k.id
+
+      UNION ALL
+
+      SELECT
+        t.id as id,
+        COUNT(a) as totalMatch,
+        (SELECT SIZE(j2.arvot) FROM t.jakaumat j2
+         WHERE j2.tyyppi = :tyoJakaumaTyyppi) as totalOsaamiset,
+        'TYOMAHDOLLISUUS' as tyyppi
+      FROM Tyomahdollisuus t
+      JOIN t.jakaumat j
+      JOIN j.arvot a
+      WHERE t.aktiivinen IS true AND j.tyyppi = :tyoJakaumaTyyppi
+      AND a.arvo IN :missingOsaamiset
+      GROUP BY t.id
```

Comment:
```suggestion
      GROUP BY t.id, j.id
```
Adding the j.id is required because the subquery generated by size(..) references the jakauma primary key. Since there is only one Osaamiset jakamauma (1:1 cardinality), this does not actually change the grouping in any way.